lib/migrator/index.ts

```diff
@@ -0,0 +1,108 @@
+import * as process from "process";
+
+import * as checkDiskSpace from 'check-disk-space';
+
+import * as permissions from "../shared/permissions";
+import winCommands from "./windows-commands";
+import * as diskpart from "../diskpart";
+import { File } from "../source-destination";
+import {
+	copyPartitionFromImageToDevice,
+	getTargetBlockDevice,
+	findPartitionsDiff
+} from "./helpers";
+
+const FREE_SPACE_NEEDED_MB = 5120
+
+/**
+ * @summary Prepares a computer running windows for switchin to balenaOS
+ * @param {string} imagePath - the balena flasher image to use as source
+ * @param {string} windowsPartition - the partition label of the device where we want to add the new data; defauls to "C"
+ * @param {string} deviceName - the storage device name, default: '\\.\PhysicalDrive0'
+ * @param {string} efiLabel - the label to use when mounting the efi partition, in case the default "M" is already in use
+ * @returns
+ */
+export const migrate = async (
+	imagePath: string, // TODO ship image with package or read from URL
+	windowsPartition: string = "C",
+	deviceName: string = "\\\\.\\PhysicalDrive0",
+	efiLabel: string = "M"
+) => {
+	console.log(`Migrate ${deviceName} with image ${imagePath}`);
+	try {
+		if (await permissions.isElevated()) {
+			if (process.platform === "win32") {
+				const freeSpace = await checkDiskSpace(`${windowsPartition}:\\`) // works only with capital
+				if ((freeSpace.free / 1024) / 1024 < FREE_SPACE_NEEDED_MB) {
+					throw Error(`Need at least ${FREE_SPACE_NEEDED_MB} MB free on partition ${windowsPartition}`)
+				}
+
+				const BOOT_FILES_SOURCE_PATH = "/EFI/BOOT";
+				const source = new File({ path: imagePath }); // will be source for next steps
+				const targetDevice = await getTargetBlockDevice(windowsPartition) // and the target
+				const originalPartitions = await targetDevice.getPartitionTable()
+
+				// step 1 make space and create partitons
+				console.log(`Shrink partition ${windowsPartition} by ${FREE_SPACE_NEEDED_MB} MB`);
+				await diskpart.shrinkPartition(windowsPartition, FREE_SPACE_NEEDED_MB);
+
+				// step 1.1
+				console.log("Create flasherBootPartition");
+				await diskpart.createPartition(deviceName, 47);
+				const afterFirstPartitions = await targetDevice.getPartitionTable()
+				const firstNewPartition = findPartitionsDiff(originalPartitions, afterFirstPartitions);
+				if (firstNewPartition.length !== 1) {
+					throw Error(`Can't query the 1. new partititon expected 1 but got ${firstNewPartition.length}`)
```

Comment:
I don't understand this error message. How about: "Expected creation of exactly 1 partition for flasher boot, but found ${firstNewPartition.length} new partitions". Need a similar message below for flasher rootA partition.

Reply:
Your poetry is better than mine for sure, go on and update it as you feel the best.
What I wanted to signal is:
1. we attempted to create a partition
2. it may have succeeded or failed
3. when looking for a single new partition we want to use later we found 0 or more than one so we don't know how to continue